packages/runtime/container-runtime/src/dataStores.ts

```diff
@@ -649,21 +649,37 @@ export class DataStores implements IDisposable {
      * This is called to update objects whose routes are unused. The unused objects are deleted.
      * @param unusedRoutes - The routes that are unused in all data stores in this Container.
      */
-    public updateUnusedRoutes(unusedRoutes: string[]) {
+    public updateUnusedRoutes(unusedRoutes: string[]): string[] {
+        const deletedRoutes = new Set<string>();
+
         for (const route of unusedRoutes) {
             const pathParts = route.split("/");
-            // Delete data store only if its route (/datastoreId) is in unusedRoutes. We don't want to delete a data
-            // store based on its DDS being unused.
+            const dataStoreId = pathParts[1];
+
+            // Skip any routes already deleted
+            if (deletedRoutes.has(route)) {
+                continue;
+            }
+
+            // Push all deleted DataStore (/datastoreId) and sub DataStore (/datastoreId/...) routes to deleted routes
+            deletedRoutes.add(route);
+
+            // Don't delete duplicate a datastore twice and don't delete a dataStore if just deleting its sub datastore
```

Comment:
nit: This sentence needs to be fixed.
Also, I wouldn't say "don't delete a dataStore if just deleting its sub datastore" but rather something along the lines of ignore sub-data store routes because a data store and its sub-routes are deleted together, so, we only need to delete the data store.

Reply:
So if we are passed a deleted dds, but not its parent, we should throw an error?

Reply:
Moving this to the validation PR